tool/tctl/common/loginrule/command.go

```diff
@@ -0,0 +1,215 @@
+	for _, subcommand := range t.subcommands {
+		match, err = subcommand.tryRun(ctx, selectedCommand, c)
+		if err != nil {
+			return match, trace.Wrap(err)
+		}
+		if match {
+			return match, nil
+		}
+	}
+	return false, nil
+}
+
+// testCommand implements the "tctl login_rule test" command.
+type testCommand struct {
+	cmd *kingpin.CmdClause
+
+	log utils.Logger
+
+	inputResourceFiles []string
+	loadFromCluster    bool
+	inputTraitsFile    string
+	outputFormat       string
+}
+
+func (t *testCommand) initialize(parent *kingpin.CmdClause, cfg *servicecfg.Config) {
+	t.cmd = parent.Command("test", "Test the parsing and evaluation of login rules.")
+	t.cmd.Flag("resource-file", "login rule resource file name (YAML or JSON)").StringsVar(&t.inputResourceFiles)
+	t.cmd.Flag("load-from-cluster", "load existing login rules from the connected Teleport cluster").BoolVar(&t.loadFromCluster)
+	t.cmd.Flag("format", "Output format: 'yaml' or 'json'").Default(teleport.YAML).StringVar(&t.outputFormat)
+	t.cmd.Arg("traits-file", "input user traits file name (YAML or JSON), empty for stdin").StringVar(&t.inputTraitsFile)
+
+	// Hack: use Alias to include some examples in the help output. This is also
+	// done elsewhere in the codebase.
+	t.cmd.Alias(`
+Examples:
+
+  Test evaluation of the login rules from rule1.yaml and rule2.yaml with input traits from traits.json
+
+  > tctl login_rule test --resource-file rule1.yaml --resource-file rule2.yaml traits.json
+
+  Test the login rule in rule.yaml along with all login rules already present in the cluster
+
+  > tctl login_rule test --resource-file rule.yaml --load-from-cluster traits.json
+
+  Read the input traits from stdin
+
+  > echo '{"groups": ["example"]}' | tctl login_rule test --resource-file rule.yaml`)
+
+	t.log = cfg.Log
+}
+
+func (t *testCommand) tryRun(ctx context.Context, selectedCommand string, c auth.ClientI) (match bool, err error) {
+	if selectedCommand != t.cmd.FullCommand() {
+		return false, nil
+	}
+
+	if len(t.inputResourceFiles) == 0 && !t.loadFromCluster {
+		return true, trace.BadParameter("no login rules to test, --resource-file or --load-from-cluster must be set")
+	}
+
+	return true, trace.Wrap(t.run(ctx, c))
+}
+
+func (t *testCommand) run(ctx context.Context, c auth.ClientI) error {
+	loginRules, err := parseLoginRuleFiles(t.inputResourceFiles)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if len(loginRules) == 0 && !t.loadFromCluster {
+		return trace.BadParameter("no login rules to test")
+	}
+
+	t.log.Infof("Loaded %d login rules from input resource files", len(loginRules))
+
+	traits, err := parseTraitsFile(t.inputTraitsFile)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	result, err := c.LoginRuleClient().TestLoginRule(ctx, &loginrulepb.TestLoginRuleRequest{
+		LoginRules:      loginRules,
+		Traits:          traitsMapResourceToProto(traits),
+		LoadFromCluster: t.loadFromCluster,
+	})
+	if err != nil {
```

Comment:
Can we do a special check for "not implemented" and make sure the error says something about the client being too new?